### PR TITLE
feat: add shutdown for frontend

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add `shutdown` method for frontend.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1158

### CODE REVIEW

File: src/frontend/src/server.rs [hunk 1]
```diff
@@ -41,19 +41,26 @@ use crate::prometheus::PrometheusOptions;
 
 pub(crate) struct Services;
 
+const GRPC_SERVER: &str = "GRPC_SERVER";
+const MYSQL_SERVER: &str = "MYSQL_SERVER";
+const POSTGRES_SERVER: &str = "POSTGRES_SERVER";
+const OPENTSDB_SERVER: &str = "OPENTSDB_SERVER";
+const HTTP_SERVER: &str = "HTTP_SERVER";
+const PROM_SERVER: &str = "PROM_SERVER";
```
Comment (MichaelScofield) [inline src/frontend/src/server.rs:49]: Can you make these names returned by the `Server` trait? I think that makes codes more concise.

File: src/frontend/src/instance.rs [hunk 2]
```diff
@@ -91,6 +94,8 @@ pub trait FrontendInstance:
 
 pub type FrontendInstanceRef = Arc<dyn FrontendInstance>;
 
+pub type ServerHandlers = HashMap<String, (Box<dyn Server>, SocketAddr)>;
```
Comment (killme2008) [inline src/frontend/src/instance.rs:97]: It's better to add a struct to replace `Box<dyn Server>, SocketAddr)`.

File: src/frontend/src/instance.rs [hunk 3]
```diff
@@ -370,12 +391,25 @@ impl Instance {
     pub fn plugins(&self) -> Arc<Plugins> {
         self.plugins.clone()
     }
+
+    pub async fn shutdown(&self) -> Result<()> {
+        futures::future::try_join_all(self.servers.values().map(|server| server.0.shutdown()))
+            .await
+            .context(error::ShutdownServerSnafu)?;
+
+        Ok(())
```
Comment (killme2008) [inline src/frontend/src/instance.rs:400]:
```suggestion
        futures::future::try_join_all(self.servers.values().map(|server| server.0.shutdown()))
            .await
            .context(error::ShutdownServerSnafu)
```

File: src/frontend/src/instance.rs [hunk 4]
```diff
@@ -370,12 +391,25 @@ impl Instance {
...
 #[async_trait]
 impl FrontendInstance for Instance {
     async fn start(&mut self) -> Result<()> {
         // TODO(hl): Frontend init should move to here
+
+        futures::future::try_join_all(self.servers.values().map(start_server))
+            .await
+            .context(error::StartServerSnafu)?;
+
         Ok(())
```
Comment (killme2008) [inline src/frontend/src/instance.rs:413]:
```suggestion
        futures::future::try_join_all(self.servers.values().map(start_server))
            .await
            .context(error::StartServerSnafu)
```

File: src/frontend/src/server.rs [hunk 5]
```diff
@@ -41,19 +41,26 @@ use crate::prometheus::PrometheusOptions;
...
         opts: &FrontendOptions,
         instance: Arc<T>,
         plugins: Arc<Plugins>,
-    ) -> Result<()>
+    ) -> Result<HashMap<String, (Box<dyn Server>, SocketAddr)>>
     where
         T: FrontendInstance,
     {
-        info!("Starting frontend servers");
+        let mut result = HashMap::<String, (Box<dyn Server>, SocketAddr)>::new();
```
Comment (killme2008) [inline src/frontend/src/server.rs:60]:
```suggestion
        let mut result = HashMap::<String, (Box<dyn Server>, SocketAddr)>::with_capacity(plugins.len());
```

File: src/frontend/src/server.rs [hunk 6]
```diff
@@ -178,48 +177,31 @@ impl Services {
...
+    server_and_addr: &(Box<dyn Server>, SocketAddr),
 ) -> servers::error::Result<Option<SocketAddr>> {
-    if let Some((server, addr)) = server_and_addr {
-        info!("Starting server at {}", addr);
-        server.start(addr).await.map(Some)
-    } else {
-        Ok(None)
-    }
+    let (server, addr) = server_and_addr;
+    info!("Starting server at {}", addr);
```
Comment (killme2008) [inline src/frontend/src/server.rs:205]: If we move the server name to the `Server` trait, then we can print the server name to log too.

File: src/frontend/src/instance.rs (hunk 4 above)
Comment (WenyXu) [reply src/frontend/src/instance.rs:413]: > Suggested change
>         futures::future::try_join_all(self.servers.values().map(start_server))
>             .await
>             .context(error::StartServerSnafu)?;
>         Ok(())
>         futures::future::try_join_all(self.servers.values().map(start_server))
>             .await
>             .context(error::StartServerSnafu)

Removing the `Ok(())` will lead to a type error. or mapping `Vec<()>` to `()`  ?
```rust
        futures::future::try_join_all(self.servers.values().map(start_server))
            .await
            .context(error::StartServerSnafu)
            .map(|_| ())
```

Comment (killme2008) [reply src/frontend/src/instance.rs:413]: Yep. It's my mistake. Ignore it.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**